Automattic/woocommerce-payments
# feat: tokenized cart ECE base implementation

Part of https://github.com/Automattic/woocommerce-payments/issues/9722

#### Changes proposed in this Pull Request









Leveraging the Store API for ECE buttons on shortcode cart & checkout.
This is the minimum amount of changes I can introduce to have a working prototype 😄 

Due to these changes, the pay-for-order, the product pages, and the block-based cart & checkout might not be working as intended.
If I were to fix that behavior, we'd need more file changes 😄 this is an iterative approach to make the tokenized cart feature work.

Most of the logic is located in `client/tokenized-express-checkout/event-handlers.js`. Here, instead of using the methods on the `WCPayAPI` to make requests to the store, we're using the `ExpressCheckoutCartApi` or `ExpressCheckoutOrderApi` classes to communicate via the Store API.
The `ExpressCheckoutCartApi` and `ExpressCheckoutOrderApi` classes have been copied from the `client/tokenized-payment-request` directory for repurpose.

The files from the `client/tokenized-payment-request` are being repurposed. Since that directory doesn't contain any files that are used anywhere in production, I am ignoring its tests and and I am ignoring any lint rules.
This way, I can better showcase which files I am moving from `client/tokenized-payment-request` to `client/tokenized-express-checkout`.
I'm not concerned with having that directory in a non-working state. My main concern is to show the files being repurposed and keeping their git history.

I also moved the AJAX handlers from `includes/class-wc-payments-payment-request-button-handler.php` into `includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php`, and created some tests for them in `tests/unit/express-checkout/test-class-wc-payments-express-checkout-ajax-handler.php`.

#### Testing instructions





- Update your WooPayments Dev Tools
- Enable the "Enable Cart-Token implementation for ECE" feature flag
- Restart your webpack build (`npm run start`)
- As a merchant, ensure you have GooglePay/ApplePay enabled in the WooPayments settings (for product pages, cart, checkout)
- As a customer, go to a product page for a simple product (virtual or physical, doesn't matter which)
- Add the product to the cart
- Navigate to the block-based cart (or checkout) pages
- You should see the GooglePay button in Chrome
- You should be able to update the cart quantities, and the totals should be reflected in GooglePay/ApplePay
- You should be able to interact with shipping addresses & shipping options (in case of a physical product)
![2024-11-20 16 41 46](https://github.com/user-attachments/assets/70d2a993-2b74-4bab-9749-8f18ee5683ca)

- When interacting with shipping options, you should see the network requests being made to your store's Store API (e.g.: `wp-json/wc/store/v1/cart/update-customer`
- The "Order detail" dialog (the ℹ️ info icon next to the price) should show the elements in the cart
<img width="468" alt="Screenshot 2024-11-20 at 4 56 24 PM" src="https://github.com/user-attachments/assets/15790103-fb52-4fe8-95c6-ab37239db5b8">

- You should be able to place the order
- The order page in `wp-admin` should reflect the cart contents, the currency being used, and the totals
<img width="301" alt="Screenshot 2024-11-20 at 4 43 29 PM" src="https://github.com/user-attachments/assets/db4f9863-f332-4b4e-9c01-efb3f571cf65">

-------------------


**Post merge**
